op-challenger/game/keccak/merkle/tree_test.go

```diff
@@ -0,0 +1,214 @@
+package merkle
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+	"testing"
+
+	"github.com/ethereum-optimism/optimism/op-challenger/game/keccak/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/stretchr/testify/require"
+)
+
+func TestBinaryMerkleTree_AddLeaf(t *testing.T) {
+	var tests []struct {
+		name      string
+		leafCount int
+	}
+
+	// Test only three leaf counts since this requires adding a lot of leaves.
+	// To test more thoroughly, increase the divisor and run locally.
+	for i := 0; i < MaxLeafCount; i += MaxLeafCount / 3 {
+		tests = append(tests, struct {
+			name      string
+			leafCount int
+		}{
+			name:      fmt.Sprintf("AddLeaf-%d", i),
+			leafCount: i,
+		})
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			tree := NewBinaryMerkleTree()
+			expectedLeafHash := zeroHashes[BinaryMerkleTreeDepth-1]
+			for i := 0; i < test.leafCount; i++ {
+				input := ([136]byte)(bytes.Repeat([]byte{byte(i)}, 136))
+				lastLeaf := types.Leaf{
+					Input:           input,
+					Index:           big.NewInt(int64(i)),
+					StateCommitment: common.Hash{},
+				}
+				tree.AddLeaf(lastLeaf)
+				expectedLeafHash = lastLeaf.Hash()
+			}
+			leaf := tree.walkDownToLeafCount(tree.LeafCount)
+			require.Equal(t, expectedLeafHash, leaf.Label)
+		})
+	}
+}
+
+func TestBinaryMerkleTree_RootHash(t *testing.T) {
+	tests := []struct {
+		name      string
+		leafCount int
+		rootHash  common.Hash
+	}{
+		{
+			name:      "EmptyBinaryMerkleTree",
+			leafCount: 0,
+			rootHash:  common.HexToHash("2733e50f526ec2fa19a22b31e8ed50f23cd1fdf94c9154ed3a7609a2f1ff981f"),
+		},
+		{
+			name:      "SingleLeaf",
+			leafCount: 1,
+			rootHash:  common.HexToHash("b726b462a50ed30781da82e5293b8c9a35279cb6b6cfe8c3e54ece15a4fbd328"),
+		},
+		{
+			name:      "TwoLeaves",
+			leafCount: 2,
+			rootHash:  common.HexToHash("96a031b4cb1dbe70a20d578ec7ab3c43d3ebfa043e64ca517021aaad7a691817"),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			tree := NewBinaryMerkleTree()
+			for i := 0; i < test.leafCount; i++ {
+				input := ([136]byte)(bytes.Repeat([]byte{0x01}, 136))
```

Comment:
Probably better to repeat the leaf index rather than having the same value for every leaf.